### DISCUSSION TITLE
Problem with x-for and HTMX

### DISCUSSION THREAD

**Discussion by florian-sun** [alpinejs/alpine#4476] [1. Help] [answered]
I have a serious problem when using x-for together with htmx.
If a have the htmx attribute hx-boost enabled, when I click the browser back button the elements in the x-for template are duplicated.

Here is my code. For simplicity I'm using a list (x-data) with one element

```
<!DOCTYPE html>
<html>
<head>
	<script src = "https://unpkg.com/htmx.org@2.0.3" integrity="sha384-0895/pl2MU10Hqc6jd4RvrthNlDiE9U1tWmX7WRESftEDRosgxNsQG/Ze9YMRzHq" crossorigin="anonymous"></script>
	<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
</head>
<body>   

  <div  hx-boost="true"> 
    <a  href="/login"> Login </a>

    <template  x-data="{ items: ['foo'] }" x-for="item in items" :key="item"> 
      <div > 
        <p  x-text="item"> </p>
      </div>
    </template>
  </div>

</body>
</html>

```
The rendered page is this (only the body content):
```

<div hx-boost="true"> 
  <a href="/login"> Login </a>

  <template x-data="{ items: ['foo'] }" x-for="item in items" :key="item" > 
    <div> 
      <p x-text="item"> </p>
    </div>
   </template>
   <div> 
     <p x-text="item">foo</p>
   </div>
</div>
```

After clicking login link and then the back button, the rendered page is this:

```
<div hx-boost="true"> 
  <a href="/login"> Login</a>

  <template x-data="{ items: ['foo'] }" x-for="item in items" :key="item"> 
    <div> 
      <p x-text="item"> </p>
    </div>
  </template>
  <div> 
    <p x-text="item">foo</p>
  </div>
  <div> 
    <p x-text="item"></p>
  </div>
</div>

```

As you can see, a new div was added. Every time the back button is clicked, a new div appear.
In this case it's not much damage because the duplicated content is empty. However usually I have more then one element in the root div and all of them are duplicated.

Browser is Edge.
The login page is this

```
<body >
<p> login page</p>
</body>
```

**Comment by SimoTod** [comment]
does it work if you use https://v1.htmx.org/extensions/alpine-morph/?

**Comment by florian-sun** [comment]
Not sure how this helps. Does the alpine morph replace htmx swapping mechanism so I don't need to use boost anymore?

**Reply by SimoTod** [reply]
No, you still need to use boost. The internal swapping algorithm used by htmx is not fully compatible with Alpine because it removes all the hidden metadata added by Alpine, that addon just change it so Htmx uses alpine morph instead.

**Comment by florian-sun** [comment]
Thanks, I tried but no luck.
Actually what I did was just installing the extension. As you see in my example, I am not using any swapping or other htmx stuff, I'm just adding the attribute and that alone is enough to cause this problem.

**Reply by SimoTod** [reply]
I think hx-boost internally swaps the html. it does an ajax call to fetch the new content and swap it in your page without fully reloading the rest. I've tried anyway and it doesn't help as you said, when you click back htmx is restoring the html from local storage, which also contains the html generated by x-for but when inserted in page in that way, alpine cannot link it correctly anymore.

As a workaround, add `hx-history="false"` on the parent element inside x-for so htmx will not cache it (you probably have to empty the local storage).

```html
<!DOCTYPE html>
<html>
<head>
        <script src = "https://unpkg.com/htmx.org@2.0.3" integrity="sha384-0895/pl2MU10Hqc6jd4RvrthNlDiE9U1tWmX7WRESftEDRosgxNsQG/Ze9YMRzHq" crossorigin="anonymous"></script>
        <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
</head>
<body>

  <div hx-boost="true">
    <a href="/login.html"> Login </a>

    <template x-data="{ items: ['foo'] }" x-for="item in items" :key="item">
      <div hx-history="false">
        <p x-text="item"> </p>
      </div>
    </template>
  </div>

</body>
</html>
```

**Reply by SimoTod** [reply]
i suggest you add it to any x-for and also x-if as I suspect that directive is going to have the same issue if used with HTMX

**Reply by SimoTod** [reply]
In theory, you can also write a plugin for htmx that hooks into the htmx:beforeHistorySave event and remove the content generated from x-for and x-if from the cache snapshot (more performant then hx-history="false" because the latter will just disable the cache for that page).
I wrote something similar for turbolinks a while ago

**Reply by SimoTod** [reply, marked as answer]
This is a simple snippet of code to fix x-for and x-if when used with the HTMX history
```html
    ...
    <script src = "https://unpkg.com/htmx.org@2.0.3" integrity="sha384-0895/pl2MU10Hqc6jd4RvrthNlDiE9U1tWmX7WRESftEDRosgxNsQG/Ze9YMRzHq" crossorigin="anonymous"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
    <script>
        document.addEventListener('htmx:beforeHistorySave', (evt) => {
            document.querySelectorAll('[x-for]').forEach((item) => {
                item._x_lookup && Object.values(item._x_lookup).forEach((el) => el.remove())
            })
            document.querySelectorAll('[x-if]').forEach((item) => {
                item._x_currentIfEl && item._x_currentIfEl.remove()
            })
        })
    </script>
    ...
```

**Reply by florian-sun** [reply]
Yes, this did the trick.
Also the one you showed before with hx-history="false" works but not completely. I must check also for the header HX-History-Restore-Request and load the full page when it happens. This one is much better.
Thanks a lot!

**Reply by believelody** [reply]
> This is a simple snippet of code to fix x-for and x-if when used with the HTMX history
> 
> ```
>     ...
>     <script src = "https://unpkg.com/htmx.org@2.0.3" integrity="sha384-0895/pl2MU10Hqc6jd4RvrthNlDiE9U1tWmX7WRESftEDRosgxNsQG/Ze9YMRzHq" crossorigin="anonymous"></script>
>     <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
>     <script>
>         document.addEventListener('htmx:beforeHistorySave', (evt) => {
>             document.querySelectorAll('[x-for]').forEach((item) => {
>                 item._x_lookup && Object.values(item._x_lookup).forEach((el) => el.remove())
>             })
>             document.querySelectorAll('[x-if]').forEach((item) => {
>                 item._x_currentIfEl && item._x_currentIfEl.remove()
>             })
>         })
>     </script>
>     ...
> ```

We need to make it as an extension or whatever we can do !!! Save a lot of despairing time lol. Same thing for x-if has to be done as well. Thanks a lot @SimoTod 🙏🏿

**Reply by sacredscarcity** [reply]
Thanks for this @SimoTod. the snippet you provided fixed the x-for and x-if but for x-model it still lost the state.

**Reply by SimoTod** [reply]
hi @sacredscarcity this snippet doesn't address the alpine state, it only stops DOM elements generated by alpine going into the htmx cache which would cause an error when htmx restores the page (when you click the back button).
To preserve alpine state when htmx swaps element you need to look into their extension: https://github.com/bigskysoftware/htmx-extensions/blob/main/src/alpine-morph/README.md

**Reply by sacredscarcity** [reply]
Thank you for pointing me to the right direction!
